server/src/main/java/org/opensearch/indices/replication/SegmentReplicationSourceService.java

```diff
@@ -172,7 +173,11 @@ public void clusterChanged(ClusterChangedEvent event) {
                 for (IndexShard indexShard : indexService) {
                     if (indexShard.routingEntry().primary()) {
                         final IndexMetadata indexMetadata = indexService.getIndexSettings().getIndexMetadata();
-                        final Set<String> inSyncAllocationIds = indexMetadata.inSyncAllocationIds(indexShard.shardId().id());
+                        final Set<String> inSyncAllocationIds = new HashSet<>(indexMetadata.inSyncAllocationIds(indexShard.shardId().id()));
+                        if (indexShard.isPrimaryMode()) {
+                            final Set<String> shardTrackerInSyncIds = indexShard.getReplicationGroup().getInSyncAllocationIds();
+                            inSyncAllocationIds.addAll(shardTrackerInSyncIds);
```

Comment:
do we need both sets of ids?

Reply:
Good point. 

Using only cluster state sync ids was not enough (this bug).
Using only tracker's sync ids may result in early cancellations before actual cluster state update. This might be problematic when actual update (removal from insync set) never reaches cluster-manager. By waiting for cluster state update, we are being less aggressive towards cancellations, which I feel is better.